### DISCUSSION TITLE
Allow two GPIOs share same PWM

### DISCUSSION THREAD

**Discussion by GagoX** [arendst/Tasmota#10884] [Q&A] [answered]
Hello,
In the menu "Configure Module", it would be useful to be able to assign one PWM output to two different GPIOs.

For example, when using the Shelly RGBW2 (4 outputs), one could assign PWM1 to two different GPIOs and PWM2 to the other two GPIOs. This way tasmota would work as a 2Channel CCT Light controller and one could use all the power available from the Shelly device and control two or more CCT lights in parallel. 

I tried this using Tasmota v9.2.0 but it did not work: only oneGPIO outputs a PWM signal. Look for it in the documentations but could not find it. 

May be there is a way to mirror two GPIOs directly in the code or through a SetOption command?

What do you think?

Thanks and Regards

**Comment by sfromis** [comment, marked as answer]
One way could be to split the PWM channels into up to 5 separate lights using SetOption68, and then do the same commands to both PWM channels in a pair.

**Reply by barbudor** [reply]
On top of Sfromis answer, using DeviceGroups it should be possible to pair multiple PWM together
If not, Rules are your friend

**Reply by GagoX** [reply]
HI sfromis, hi barbudor, 
thanks for your suggestions. The thing is that when setting Tasmota to use PWM1 and PWM2 only, I can use the CCT and DIMMER commands which in turn configure the PWM1 and PWM2 duty cycles automatically. 
I thought there may be a way to have tasmota as 2Ch CCT light with 2x2 outputs.
I will look into defining rules to mimic the CCT light behaviour. I already have some rather complex rules that I use for controlling Dimming and CCT (and presets) with one switch. Ill see what I can do. Thanks!

**Reply by sfromis** [reply]
You could simply wire same gpio to both led driver transistors. Not tested this, but I'd expect it to work.

**Reply by GagoX** [reply]
Hi sfromis, yes, it would work but this would mean to mod the Shelly PCB. For DIY solutions it would not be a problem of course

**Reply by sfromis** [reply]
Of course, that is exactly what it would mean. Not a "problem", it is *your* device.

**Reply by GagoX** [reply]
Thanks for the ideas.

I do not yet understand how DeviceGroups would work... I thought this feature was meant to group multiple tasmota devices and do MQTT control. 

For the moment I will move forward with the workaround of  going with SetOption68, Rules and two Mem<x> variables to calculate the Channel<X>. The Mem1 and Mem2 shall represent the Dimmer and CT values. 

The drawback of this approach are:
 - I will not be able to use the web interface with the sliders to set the Dimmer and CT and use all the advantages of SetOption15=1. 
 - Another drawback is that my rules will get bulkier and slower. (I use rules to control Dimmer and CT and do some presets using a single push button configured as switch with SwitchOption11).
 - For controlling additionally with  Home Assistant I will have to figure out how to change the value of the Mem<x> variable over MQTT and add more rule  triggers in Tasmota to calculate the Channel<x>. This is also not a standard solution in HA.  

All in all I still think that allowing two GPIOs share the same PWM component would be a more straightforward solution, at least from the user point of view. Of course I am not fully aware of all the technical implications in the Tasmota code.

Do you think the original idea might see the light of day? 
Regards

**Reply by arendst** [reply]
Alas I don't think so. The proposed solution of connecting two strings to the same output is the most viable solution.

**Reply by GagoX** [reply]
got it ;)

**Comment by NoobyNiceDev** [comment]
Hey!
I also tried Tasmota on a Shelly RGBW2 and I am very happy. Just that one thing, that I cannot use all 4 channels of the shelly to control 2 CCT LED stripes (with identical PWM and color-ratio). Shelly can switch up to 288W and 90W per Channel. I have to control ~450W LED stripes, so would be nice to that with two shellys, not three...
Please add the support of "mirroring" the PWM to the other 2 Outputs. Or I have to fork :)

**Reply by s-hadinger** [reply]
This is a edge case. If you find a developer motivated to implement it that's doable. You could use the fact that you can assign two identical pwm to 2 different GPIOs. Although Tasmota is not designed to handle such configuration. 

You are welcome to provide a Pull Request.

**Comment by sudden6** [comment]
Hi, I'm having the same problem, I have a light where there are 2 pins going to a WW and CW led strip, and two other pins go to a different WW and CW led strip. It would be nice if tasmota could provide 2 sets of sliders for that setup to control each zone separately.

For now it would help if there was a way to set both zones of the light to the same settings, is there a way to do that without loosing the WebUI?

I'd also be willing to dive into the code a bit, but after a quick scan I can't find where the mapping of PWM channel to GPIO pin is done. Can someone give me a hint?

**Reply by s-hadinger** [reply]
The number of possible combinations is already too high. I'm afraid there is no simple way unless you want to code it in Berry on top of the core light. Berry is only on esp32 variantw

**Reply by sfromis** [reply]
One "partial" approach could be to config 5 PWM channels, one of the first 3 on an unused pin, and using `ChannelRemap 128` to split the RGBCW setup into independent channels for RGB and CW. If you then use the red+blue channels (with green having no effect), selecting colors which are a mix of red/blue would become mixes of cold/warm white, resulting in varying color temperatures.

On Shelly RGBW2, this would mean configuring one of R/G/B on an unused gpio to get to 5 channels, of which you could use terminals R/B for one light, where you control the mixing, and G/W as CW/WW for another.

**Reply by sudden6** [reply]
@s-hadinger Even for the simple variant, where the same pwm value is just output on two GPIOs? AFAICT this is already selectable in the WebUI, e.g. if GPIO 17 and 18 can be assigned as PWM 1, but only GPIO 17 actually outputs the PWM signal.

I'm only using ESP32 btw, so for me personally it wouldn't be a problem if this doesn't work on the ESP8266.

Can you point me to the souce code where this light/PWM thing is implemented? If all else fails I'll probably just hack it in for the light I have.

**Reply by s-hadinger** [reply]
For esp32 it's here but it will not be a simple change
https://github.com/arendst/Tasmota/blob/d29b3b2871905a0670f372f4c7e1772d1dcb4928/tasmota/tasmota_support/support_pwm.ino#L78

**Comment by sudden6** [comment]
I have devoloped a patch that fixes the problem for my use case and opened a PR here: https://github.com/arendst/Tasmota/pull/20760 feel free to use and test at your own risk.
